python/ray/serve/tests/test_cli.py

```diff
@@ -27,5 +31,40 @@ def test_start_shutdown_in_namespace(ray_start_stop):
     subprocess.check_output(["serve", "-n", "test", "shutdown"])
 
 
+@serve.deployment
+class A:
+    def __init__(self, value, increment=1):
+        self.value = value
+        self.increment = increment
+        self.decrement = 0
+
+    def reconfigure(self, config):
+        self.decrement = config["decrement"]
+
+    def __call__(self, inp):
+        return self.value + self.increment - self.decrement
+
+
+def test_deploy(ray_start_stop):
+    subprocess.check_output(["serve", "start"])
+    subprocess.check_output([
+        "serve", "deploy", "ray.serve.tests.test_cli.A", "--options-json",
```

Comment:
can we test with a working_dir and import from a file in it.

Reply:
hmmm @edoakes `working_dir` in `runtime_env` dict unfortunately doesn't work.


```
(ServeController pid=25773) During handling of the above exception, another exception occurred:
(ServeController pid=25773) 
(ServeController pid=25773) Traceback (most recent call last):
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/controller.py", line 199, in run_control_loop
(ServeController pid=25773)     self.deployment_state_manager.update()
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/deployment_state.py", line 1388, in update
(ServeController pid=25773)     deleted = deployment_state.update()
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/deployment_state.py", line 1141, in update
(ServeController pid=25773)     running_replicas_changed = self._scale_deployment_replicas()
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/deployment_state.py", line 919, in _scale_deployment_replicas
(ServeController pid=25773)     self._target_version)
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/deployment_state.py", line 395, in start
(ServeController pid=25773)     self._actor.start(deployment_info, version)
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/deployment_state.py", line 189, in start
(ServeController pid=25773)     **deployment_info.replica_config.ray_actor_options).remote(
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/actor.py", line 496, in options
(ServeController pid=25773)     new_runtime_env = ParsedRuntimeEnv(runtime_env)
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/_private/runtime_env/validation.py", line 299, in __init__
(ServeController pid=25773)     validated_option_val = validate_fn(option_val)
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/_private/runtime_env/validation.py", line 73, in parse_and_validate_working_dir
(ServeController pid=25773)     validate_uri(working_dir)
(ServeController pid=25773)   File "/Users/simonmo/Desktop/ray/ray/python/ray/_private/runtime_env/validation.py", line 39, in validate_uri
(ServeController pid=25773)     f"{uri} is not a valid URI. Passing directories or modules to "
(ServeController pid=25773) ValueError: /tmp/tmpkdcsd4cy is not a valid URI. Passing directories or modules to be dynamically uploaded is only supported at the job level (i.e., passed to `ray.init`).
^C
```
I added a --runtime-env-json to all `serve` command.

Reply:
Ah, this is because you can't do dynamic uploads at a per-actor level (this is ok).

Can you please test with HTTP/S3 instead?

Reply:
@edoakes actually let me do that in a follow up... the package_uri doesn't support http, and i tried the following but it still doesn't work yet. 

```diff
diff --git a/python/ray/_private/runtime_env/packaging.py b/python/ray/_private/runtime_env/packaging.py
index d8be1f455..18f7b69b1 100644
--- a/python/ray/_private/runtime_env/packaging.py
+++ b/python/ray/_private/runtime_env/packaging.py
@@ -40,6 +40,8 @@ class Protocol(Enum):
 
     GCS = "gcs", "For packages dynamically uploaded and managed by the GCS."
     CONDA = "conda", "For conda environments installed locally on each node."
+    HTTP = "http", ("Remote htts path, "
+                    "assumes everything packed in one zip file.")
     HTTPS = "https", ("Remote https path, "
                       "assumes everything packed in one zip file.")
     S3 = "s3", "Remote s3 path, assumes everything packed in one zip file."
@@ -50,7 +52,7 @@ class Protocol(Enum):
     def remote_protocols(cls):
         # Returns a lit of protocols that support remote storage
         # These protocols should only be used with paths that end in ".zip"
-        return [cls.HTTPS, cls.S3, cls.GS]
+        return [cls.HTTP, cls.HTTPS, cls.S3, cls.GS]
 
 
 def _xor_bytes(left: bytes, right: bytes) -> bytes:
@@ -132,18 +134,18 @@ def parse_uri(pkg_uri: str) -> Tuple[Protocol, str]:
                 netloc='_ray_pkg_029f88d5ecc55e1e4d64fc6e388fd103.zip'
             )
             -> ("gcs", "_ray_pkg_029f88d5ecc55e1e4d64fc6e388fd103.zip")
-    For HTTPS URIs, the netloc will have '.' replaced with '_', and
+    For HTTP(S) URIs, the netloc will have '.' replaced with '_', and
     the path will have '/' replaced with '_'. The package name will be the
-    adjusted path with 'https_' prepended.
+    adjusted path with 'http(s)_' prepended.
         urlparse(
-            "https://github.com/shrekris-anyscale/test_module/archive/HEAD.zip"
+            "http(s)://github.com/shrekris-anyscale/test_module/archive/HEAD.zip"
         )
             -> ParseResult(
-                scheme='https',
+                scheme='http(s)',
                 netloc='github.com',
                 path='/shrekris-anyscale/test_repo/archive/HEAD.zip'
             )
-            -> ("https",
+            -> ("http(s)",
             "github_com_shrekris-anyscale_test_repo_archive_HEAD.zip")
     For S3 URIs, the bucket and path will have '/' replaced with '_'. The
     package name will be the adjusted path with 's3_' prepended.
@@ -170,11 +172,10 @@ def parse_uri(pkg_uri: str) -> Tuple[Protocol, str]:
     if protocol == Protocol.S3 or protocol == Protocol.GS:
         return (protocol,
                 f"{protocol.value}_{uri.netloc}{uri.path.replace('/', '_')}")
-    elif protocol == Protocol.HTTPS:
-        return (
-            protocol,
-            f"https_{uri.netloc.replace('.', '_')}{uri.path.replace('/', '_')}"
-        )
+    elif protocol == Protocol.HTTPS or protocol == Protocol.HTTP:
+        return (protocol,
+                (f"{protocol.value.lower()}_{uri.netloc.replace('.', '_')}"
+                 f"{uri.path.replace('/', '_')}"))
     else:
         return (protocol, uri.netloc)
 
diff --git a/python/ray/serve/tests/test_cli.py b/python/ray/serve/tests/test_cli.py
index db43a276c..2e5af4f32 100644
--- a/python/ray/serve/tests/test_cli.py
+++ b/python/ray/serve/tests/test_cli.py
@@ -2,13 +2,17 @@ import json
 import os
 from pathlib import Path
 import subprocess
+import tempfile
+import random
 import sys
+import zipfile
 
 import pytest
 import requests
 
 from ray import serve
 from ray.tests.test_runtime_env_working_dir import tmp_working_dir  # noqa: F401, E501
+from ray._private.test_utils import wait_for_condition
 
 
 @pytest.fixture
@@ -54,7 +58,35 @@ class A:
         return (self.value + self.increment - self.decrement) * self.multiplier
 
 
+@pytest.fixture
+def http_file_server():
+    dir_path = tempfile.mkdtemp()
+    free_port = random.randint(30000, 45000)
+    proc = subprocess.Popen(
+        [
+            sys.executable,
+            "-m",
+            "http.server",
+            str(free_port),
+        ], cwd=dir_path)
+
+    # sanity check
+    def _check():
+        try:
+            resp = requests.get(f"http://127.0.0.1:{free_port}", timeout=1)
+            resp.raise_for_status()
+            return True
+        except Exception:
+            return False
+
+    wait_for_condition(_check)
+    yield dir_path, free_port
+    proc.kill()
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Fail to create temp dir.")
 def test_deploy(ray_start_stop, tmp_working_dir):  # noqa: F811
+    # Test package import path
     subprocess.check_output(["serve", "start"])
     subprocess.check_output([
         "serve", "--runtime-env-json",
@@ -84,6 +116,44 @@ def test_deploy(ray_start_stop, tmp_working_dir):  # noqa: F811
     resp.raise_for_status()
     assert resp.text == "94", resp.text
 
+    subprocess.check_output(["serve", "shutdown"])
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Fail to create temp dir.")
+def test_deploy_zip_package(ray_start_stop, http_file_server):  # noqa: F811
+    subprocess.check_output(["serve", "start"])
+
+    # Test file package path
+    dir_path, free_port = http_file_server
+    dir_path = Path(dir_path)
+    with open(dir_path / "my_module.py", "w") as f:
+        f.write(f"""
+from ray import serve
+@serve.deployment
+def f():
+    return "{free_port}"
+""")
+    with zipfile.ZipFile(dir_path / "my_repo.zip", "w") as my_zip:
+        my_zip.write(dir_path / "my_module.py")
+
+    subprocess.check_output(
+        [
+            "serve", "deploy", "my_module.f", "--options-json",
+            json.dumps({
+                "ray_actor_options": {
+                    "runtime_env": {
+                        "working_dir": f"http://127.0.0.1:{free_port}/my_repo.zip"
+                    }
+                }
+            })
+        ],
+        cwd=dir_path)
+    resp = requests.get("http://127.0.0.1:8000/f")
+    resp.raise_for_status()
+    assert resp.text == f"{free_port}", resp.text
+
+    subprocess.check_output(["serve", "shutdown"])
+
 
 if __name__ == "__main__":
     sys.exit(pytest.main(["-v", "-s", __file__]))
diff --git a/python/ray/tests/test_runtime_env_packaging.py b/python/ray/tests/test_runtime_env_packaging.py
index 67e90bb1f..2e3f00521 100644
--- a/python/ray/tests/test_runtime_env_packaging.py
+++ b/python/ray/tests/test_runtime_env_packaging.py
@@ -290,6 +290,7 @@ def test_travel():
     "parsing_tuple",
     [("gcs://file.zip", Protocol.GCS, "file.zip"),
      ("s3://bucket/file.zip", Protocol.S3, "s3_bucket_file.zip"),
+     ("http://test.com/file.zip", Protocol.HTTP, "https_test_com_file.zip"),
      ("https://test.com/file.zip", Protocol.HTTPS, "https_test_com_file.zip"),
      ("gs://bucket/file.zip", Protocol.GS, "gs_bucket_file.zip")])
 def test_parsing(parsing_tuple):
```

@shrekris-anyscale is there other things i should change? I kept getting 
```
(raylet) [2021-11-16 11:49:18,164 E 43015 14745200] agent_manager.cc:190: Failed to create runtime env: {"working_dir": "http://127.0.0.1:38689/my_repo.zip"}, error message: [Errno 2] No such file or directory: '/tmp/ray/session_2021-11-16_11-49-06_598367_42991/runtime_resources/working_dir_files/http_127_0_0_1:38689_my_repo/tmp/tmpb3vj__di/my_module.py' -> '/tmp/ray/session_2021-11-16_11-49-06_598367_42991/runtime_resources/working_dir_files/http_127_0_0_1:38689_my_repo/tmpb3vj__di/my_module.py'
```

@edoakes Additionally we don't support imported backend here yet so packaging doesn't make too much sense because you still need the file to present on the caller side.

Reply:
... actually, @architkulkarni is this file not found error related to gc?

Reply:
We haven't added support for HTTP yet. We just have support for HTTPS, so an HTTP remote URI will throw an error.

Reply:
What does the local zip file contain? I think the error might be in the `extract_file_and_remove_top_level_dir()` function in `packaging.py`. That function extracts specific files from a zip file, and then it removes them from the top level directory in the zip file. Based on the error message it looks like that function is trying to move the `/tmpb3vj__di/my_module.py` file outside of the top level `tmp` directory, but `/tmp/ray/session_2021-11-16_11-49-06_598367_42991/runtime_resources/working_dir_files/http_127_0_0_1:38689_my_repo/tmp/tmpb3vj__di/my_module.py` isn't a valid path.

The issue might be the a colon in the `http_127_0_0_1:38689_my_repo` directory name. On my computer, when I try to make a directory with a colon, the OS replaces it with `/`. Can you try replacing `{uri.netloc.replace('.', '_')}` with `{uri.netloc.replace('.', '_').replace(':', '_')}` in the `elif protocol == Protocol.HTTPS or protocol == Protocol.HTTP:` clause in `parse_uri()`? That might fix it.

Also, there's a typo in the Protocol enum's HTTP definition. The description says htts instead of http.

Reply:
thx @shrekris-anyscale, moved things over to https://github.com/ray-project/ray/pull/20518